crates/rust/src/lib.rs

```diff
@@ -307,6 +336,388 @@ impl RustWasm {
+    } else {
+        core::hint::unreachable_unchecked()
+    }
+}
+                    ",
+                );
+            }
+
+            RuntimeItem::CharLift => {
+                self.src.push_str(
+                    "\
+pub unsafe fn char_lift(val: u32) -> char {
+    if cfg!(debug_assertions) {
+        core::char::from_u32(val).unwrap()
+    } else {
+        core::char::from_u32_unchecked(val)
+    }
+}
+                    ",
+                );
+            }
+
+            RuntimeItem::BoolLift => {
+                self.src.push_str(
+                    "\
+pub unsafe fn bool_lift(val: u8) -> bool {
+    if cfg!(debug_assertions) {
+        match val {
+            0 => false,
+            1 => true,
+            _ => panic!(\"invalid bool discriminant\"),
+        }
+    } else {
+        core::mem::transmute::<u8, bool>(val)
+    }
+}
+                    ",
+                );
+            }
+
+            RuntimeItem::RunCtorsOnce => {
+                self.src.push_str(
+                    r#"
+/// Provide a hook for generated export functions to run static
+/// constructors at most once. wit-bindgen-rust generates a call to this
+/// function at the start of all component export functions. Importantly,
+/// it is not called as part of `cabi_realloc`, which is a *core* export
+/// func, but may not execute ctors, because the environment ctor in
+/// wasi-libc (before rust 1.69.0) calls an import func, which is not
+/// permitted by the Component Model when inside realloc.
+#[cfg(target_arch = "wasm32")]
+pub fn run_ctors_once() {
+    static mut RUN: bool = false;
+    unsafe {
+        if !RUN {
+            // This function is synthesized by `wasm-ld` to run all static
+            // constructors. wasm-ld will either provide an implementation
+            // of this symbol, or synthesize a wrapper around each
+            // exported function to (unconditionally) run ctors. By using
+            // this function, the linked module is opting into "manually"
+            // running ctors.
+            extern "C" {
+                fn __wasm_call_ctors();
+            }
+            __wasm_call_ctors();
+            RUN = true;
+        }
+    }
+}
+                    "#,
+                );
+            }
+
+            RuntimeItem::AsI32 => {
+                self.emit_runtime_as_trait(
+                    "i32",
+                    &["i32", "u32", "i16", "u16", "i8", "u8", "char", "usize"],
+                );
+            }
+
+            RuntimeItem::AsI64 => {
+                self.emit_runtime_as_trait("i64", &["i64", "u64"]);
+            }
+
+            RuntimeItem::AsF32 => {
+                self.emit_runtime_as_trait("f32", &["f32"]);
+            }
+
+            RuntimeItem::AsF64 => {
+                self.emit_runtime_as_trait("f64", &["f64"]);
+            }
+
+            RuntimeItem::ResourceType => {
+                self.src.push_str(
+                    r#"
+type RawRep<T> = Option<T>;
```

Comment:
At some point we might want to move this to its own .rs file and `include!` it.  Likewise any other big code literals.

Reply:
`include_str!`, I meant.